bluemira/power_cycle/base.py

```diff
@@ -0,0 +1,468 @@
+# CLASS PROPERTY DECORATOR
+# ######################################################################
+class classproperty(object):
+    def __init__(self, fget):
+        self.fget = fget
+
+    def __get__(self, owner_self, owner_cls):
+        return self.fget(owner_cls)
+
+
+# ######################################################################
+# POWER CYCLE ABSTRACT BASE CLASS
+# ######################################################################
+
+
+class PowerCycleABC(abc.ABC):
+    """
+    Abstract base class for classes in the Power Cycle module.
+
+    Parameters
+    ----------
+    name: `str`
+        Description of the instance.
+    """
+
+    # ------------------------------------------------------------------
+    # CLASS ATTRIBUTES
+    # ------------------------------------------------------------------
+
+    # Plot defaults (arguments for `matplotlib.pyplot.plot`)
+    _plot_kwargs = {
+        "c": "k",  # Line color
+        "lw": 2,  # Line width
+        "ls": "-",  # Line style
+    }
+    _scatter_kwargs = {
+        "c": "k",  # Marker color
+        "s": 100,  # Marker size
+        "marker": "x",  # Marker style
+    }
+
+    # Plot text settings (for `matplotlib.pyplot.text`)
+    _text_angle = 45  # rotation angle
+    _text_index = -1  # index of (time,data) point used for location
+
+    # ------------------------------------------------------------------
+    # CONSTRUCTOR
+    # ------------------------------------------------------------------
+
+    def __init__(self, name: str):
+
+        # Store name
+        self.name = self._validate_name(name)
+
+        # Dynamically add validation classmethod
+        # setattr(a, 'func', classmethod(func))
+
+    @classmethod
+    def _validate_name(cls, name):
+        """
+        Validate a name for class instance creation to be a string.
+        """
+        class_name = cls.__name__
+        if not isinstance(name, (str)):
+            raise TypeError(
+                f"""
+                The 'name' used to create an instance of the
+                {class_name} class must be a `str`.
+                """
+            )
+
+        return name
+
+    @classmethod
+    def _issue_error(cls, label):
+        """
+        Issue error associated with...
+        """
+
+        # Retrieve errors of that class
+        error_dict = cls._errors
+
+        # Validate error label
+        all_labels = error_dict.keys()
+        if label in all_labels:
+
+            # Retrieve particular error
+            the_error = error_dict[label]
+
+            # Retrieve error attributes
+            error_type = the_error.err_type
+            error_msg = the_error.err_msg
+
+            # Build raising function
+            raise_function = error_type + "Error"
+            raise_function = f"""
+                raise {raise_function}('''{error_msg}''')
+                """
+
+            exec(raise_function)  # Issue error
```

Comment:
Never use `exec` it is usually a security issue. This would be avoided when the errors are called where they are needed

Reply:
I understood your comment and explanation during our meeting. I will forgo of the `PowerCycleError` class and just implement errors as dict properties of each class. This way only the error message is stored in the `_errors` attribute and the `raise` command will not be dynamically called.

Reply:
I had a conversation with @ivanmaione, and I will probably create a main `PowerCycleError` class, and create subclasses for each class in the module. This way the classes do not need each an `_errors` attribute. What do you think @je-cook?

Reply:
Yep that sounds fine